### ISSUE TITLE
Using `sprig` function in `.tmpl` file for filtering systems

### ISSUE THREAD

**Issue by Noahyt** [twpayne/chezmoi#2297]
## What exactly are you trying to do?

I am trying to write a template for managing files between my personal machines and a cluster.

When I log into the cluster I am connected to any one of many login nodes whose hostname is something like `holy1234` where "1234" here may be any number. However the name always contains `holy`. Therefore I am trying to write a template filter that searches for the substring `holy`.

## What have you tried so far?

I think my problem comes from not understanding GoLang, and templates. I have tried many variations like:

```{{ .chezmoi.hostname | contains "holy" }}```

`contains` is a `sprig` function which according to the documentation should do what I want, however this command when evaluated with `execute-template` returns: "chezmoi: template: arg1:1: function "holy" not defined"

EDIT: I've now realized part of the problem was bash parsing the argument in the command line -- if I escape the inner set of quotation marks I *almost* get what I want but I can't figure out where to add the `if`.

## Where else have you checked for solutions?

* [x ] I have read [chezmoi's user guide](https://chezmoi.io/user-guide/), and not found the answer.
* [ x] I have searched [chezmoi's reference guide](https://chezmoi.io/reference/), and not found the answer.
* [ x] Other, please give details. 
* Read documentation of `sprig` and `text/template` although I had a hard time understanding the latter.

**Comment by twpayne** [comment]
You need something like:

```
{{ if (contains "holy" .chezmoi.hostname) }}
...
{{ end }}
```

To be more specific, you can use `hasPrefix` instead of `contains`.

**Comment by bradenhilton** [comment]
Isn't the issue with `execute-template` caused by how quotes in arguments are handled?

```console
❯ chezmoi --debug execute-template '{{ if (contains "<username>" .chezmoi.hostname) }}True{{ else }}False{{ end }}'
...
args=["path\\to\\chezmoi.exe","--debug","execute-template","{{ if (contains <username> .chezmoi.hostname) }}True{{ else }}False{{ end }}"]
...
```

Note the now unquoted `<username>`.

Escaping the quotes works:

```console
❯ chezmoi execute-template '{{ if (contains \"<username>\" .chezmoi.hostname) }}True{{ else }}False{{ end }}'
True
```

@twpayne Would auto escaping quotes in arguments break anything?

**Comment by twpayne** [comment]
> @twpayne Would auto escaping quotes in arguments break anything?

This depends on the shell you're using. Most (all?) UNIX shells will pass single-quoted arguments directly. It looks like the shell that you're using (PowerShell?) removes double quotes from inside single-quoted arguments unless they are escaped.

**Comment by bradenhilton** [comment]
I'm not sure that's the case, as PowerShell primarily uses backticks to escape quotes. See [here](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_quoting_rules?view=powershell-7.2#including-quote-characters-in-a-string) if interested.

TL;DR:

- Single quoted strings are verbatim strings.
- Double quoted strings are expandable strings.
- Double quotes can be escaped inside a double quoted string with a backtick.
- Consecutive quotes can also be used to escape quotes.

`"{{ if (contains \"<username>\" .chezmoi.hostname) }}True{{ else }}False{{ end }}"` in PowerShell is split into `"{{ if (contains \\"` and `"<username>\\ .chezmoi.hostname) }}True{{ else }}False{{ end }}"` in the args list.

Consecutive quotes and backtick escaped quotes also result in an unquoted `<username>`:

```console
❯ chezmoi --debug execute-template "{{ if (contains `"<username>`" .chezmoi.hostname) }}True{{ else }}False{{ end }}"
...
args=["path\\to\\chezmoi.exe","--debug","execute-template","{{ if (contains <username> .chezmoi.hostname) }}True{{ else }}False{{ end }}"]
...

❯ chezmoi --debug execute-template "{{ if (contains ""<username>"" .chezmoi.hostname) }}True{{ else }}False{{ end }}"
...
args=["path\\to\\chezmoi.exe","--debug","execute-template","{{ if (contains <username> .chezmoi.hostname) }}True{{ else }}False{{ end }}"]
...

❯ chezmoi --debug execute-template '{{ if (contains ''<username>'' .chezmoi.hostname) }}True{{ else }}False{{ end }}'
...
args=["path\\to\\chezmoi.exe","--debug","execute-template","{{ if (contains <username> .chezmoi.hostname) }}True{{ else }}False{{ end }}"]
...
```

Backslashes are used to escape quotes in programs after the arguments have been handled by PowerShell.

Now I'm even more confused about what the cause is.

**Comment by bradenhilton** [comment]
@twpayne I looked into this a bit more and the answers to [this StackOverflow question](https://stackoverflow.com/questions/6714165/powershell-stripping-double-quotes-from-command-line-arguments) indicate that it's ultimately caused by [`CommandLineToArgvW`](https://docs.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw).

CMD.exe does not support single-quoted strings, so a double-quoted string with backslash-escaped double quotes is required:

```console
❯ chezmoi execute-template "{{ if (contains \"<username>\" .chezmoi.hostname) }}True{{ else }}False{{ end }}"
True
```

Zsh works as expected with a single-quoted string containing unescaped double quotes:

```console
❯ chezmoi execute-template '{{ if (contains "<username>" .chezmoi.hostname) }}True{{ else }}False{{ end }}'
True
```

PowerShell v7.2.0+ with `Enable-ExperimentalFeature PSNativeCommandArgumentPassing` as expected with a single-quoted string containing unescaped double quotes when `$PSNativeCommandArgumentPassing` is set to either `Standard` or `Windows`:

```console
❯ $PSNativeCommandArgumentPassing = 'Standard'

❯ chezmoi execute-template '{{ if (contains "<username>" .chezmoi.hostname) }}True{{ else }}False{{ end }}'
True

❯ $PSNativeCommandArgumentPassing = 'Windows'

❯ chezmoi execute-template '{{ if (contains "<username>" .chezmoi.hostname) }}True{{ else }}False{{ end }}'
True
```

**Comment by twpayne** [comment]
Thank you very much for the investigation - very interesting (and sometimes the complexity of Windows astounds me).
